# Melbourne Housing Market

The classic Data Science challenge: Analysis and prediction of housing prices in Melbourne, Australia  
source: [Kaggle](https://www.kaggle.com/datasets/anthonypino/melbourne-housing-market)

## 0. Imports

In [60]:
import pandas as pd

In [61]:
# Set pandas options
pd.set_option('display.float_format', '{:.2f}'.format)

## 1. Data analysis

In [62]:
data = pd.read_csv('Melbourne_housing_FULL.csv')
data_raw = data.copy()

In [63]:
data_raw.shape

(34857, 21)

In [64]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

In [65]:
data_raw.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h 1480000.00      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h 1035000.00      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h 1465000.00     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016      2.50   3067.00  ...      1.00 1.00    126.00           NaN   
1  3/12/2016      2.50   3067.00  ...      1.00 1.00    202.00           NaN   
2  4/02/2016      2.50   3067.00  ...      1.00 0.00    156.00         79.00   
3  4/02/2016      2.50   3067.00  ...      2.00 1.00      0.00           NaN   
4  4/03/2017      2.50   3067.00  ...      2.00 0.00    134.00        150.00   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council    -37.80      145.00  Northern Metropolitan   
1        NaN  Yarra City Council    -37.80      145.00  Northern Metropolitan   
2    1900.00  Yarra City Council    -37.81      144.99  Northern Metropolitan   
3        NaN  Yarra City Council    -37.81      145.01  Northern Metropolitan   
4    1900.00  Yarra City Council    -37.81      144.99  Northern Metropolitan   

  Propertycount  
0       4019.00  
1       4019.00  
2       4019.00  
3       4019.00  
4       4019.00  

[5 rows x 21 columns]

In [66]:
data_raw.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

In [67]:
# Data without price is useless
data_raw.dropna(subset=['Price'], inplace=True)

In [68]:
data_raw.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price                0
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          6441
Bathroom          6447
Car               6824
Landsize          9265
BuildingArea     16591
YearBuilt        15163
CouncilArea          3
Lattitude         6254
Longtitude        6254
Regionname           3
Propertycount        3
dtype: int64

In [87]:
# Splitting of data into numerical and categorical data
data_raw_num = data_raw.select_dtypes(include=['int64', 'float64'])
data_raw_cat = data_raw.select_dtypes(include=['object'])
print(f"shape of data_raw_num: {data_raw_num.shape}")
print(f"shape of data_raw_cat: {data_raw_cat.shape}")
print(f"shape of full dataset: {data_raw.shape}")

shape of data_raw_num: (27247, 13)
shape of data_raw_cat: (27247, 8)
shape of full dataset: (27247, 21)


## 2. Data Cleaning
### 2.1 Numerical Data

In [70]:
data_raw_num.describe()

Rooms       Price  Distance  Postcode  Bedroom2  Bathroom      Car  \
count 27247.00    27247.00  27246.00  27246.00  20806.00  20800.00 20423.00   
mean      2.99  1050173.34     11.28   3113.80      3.05      1.59     1.72   
std       0.95   641467.13      6.79    111.14      0.96      0.70     0.99   
min       1.00    85000.00      0.00   3000.00      0.00      0.00     0.00   
25%       2.00   635000.00      6.40   3046.00      2.00      1.00     1.00   
50%       3.00   870000.00     10.50   3088.00      3.00      1.00     2.00   
75%       4.00  1295000.00     14.00   3153.00      4.00      2.00     2.00   
max      16.00 11200000.00     48.10   3978.00     20.00      9.00    18.00   

       Landsize  BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
count  17982.00      10656.00   12084.00   20993.00    20993.00       27244.00  
mean     593.49        156.83    1966.61     -37.81      145.00        7566.78  
std     3757.27        449.22      36.76       0.09        0.12        4492.38  
min        0.00          0.00    1196.00     -38.19      144.42          83.00  
25%      219.00        100.00    1950.00     -37.86      144.93        4294.00  
50%      512.00        133.00    1970.00     -37.80      145.00        6567.00  
75%      664.00        182.00    2000.00     -37.75      145.07       10412.00  
max   433014.00      44515.00    2019.00     -37.40      145.53       21650.00

Some observations that will need further investigation:
* The minimum of 'Bedroom2' is 0.
* The minimum of 'Bathroom' is 0.
* The minimum of 'Landsize' is 0 [meters].
* The maximum of 'Landsize' is 4330141 [meters].
* The minimum value of 'BuildingArea' is 0.
* The maximum value of 'BuildingArea' is 44515 [meters].
* The minimum value of 'YearBuilt' is 1196. (Melbourne was founded in 1835)
* The maximum value of 'YearBuilt' is 2019. (Dataset was created in 2018)

In [71]:
print(f"Number of entries with bedroom2 < 1: {(data_raw_num['Bedroom2'] < 1).sum()}")
print(f"Number of entries with bathroom < 1: {(data_raw_num['Bathroom'] < 1).sum()}")
print(f"Number of entries with landsize = 0: {(data_raw_num['Landsize'] == 0).sum()}")
print(f"Number of entries with buildingarea = 0: {(data_raw_num['BuildingArea'] == 0).sum()}")
print(f"Number of entries with YearBuilt < 1835: {(data_raw_num['YearBuilt'] < 1835).sum()}")
print(f"Number of entries with YearBuilt > 2018: {(data_raw_num['YearBuilt'] > 2018).sum()}")
print(f"Number of entries with buildingarea > landsize: {(data_raw_num['BuildingArea'] > data_raw_num['Landsize']).sum()}")

Number of entries with bedroom2 < 1: 16
Number of entries with bathroom < 1: 34
Number of entries with landsize = 0: 1942
Number of entries with buildingarea = 0: 61
Number of entries with YearBuilt < 1835: 3
Number of entries with YearBuilt > 2018: 1
Number of entries with buildingarea > landsize: 1373


In [72]:
# Drop rows with invalid data 
# (probably typos, but not that many that it would be worth further investigation)
data_raw_num.drop(data_raw_num[data_raw_num['YearBuilt'] < 1835].index, inplace=True)
data_raw_num.drop(data_raw_num[data_raw_num['YearBuilt'] > 2018].index, inplace=True)

In [73]:
# Set invalid values to NaN
data_raw_num.loc[data_raw_num['Bedroom2'] < 1, 'Bedroom2'] = None
data_raw_num.loc[data_raw_num['Bathroom'] < 1, 'Bathroom'] = None
data_raw_num.loc[data_raw_num['Landsize'] == 0, 'Landsize'] = None
data_raw_num.loc[data_raw_num['BuildingArea'] == 0, 'BuildingArea'] = None

In [80]:
print(f"Number of entries with buildingarea > landsize: {(data_raw_num['BuildingArea'] > data_raw_num['Landsize']).sum()}")

Number of entries with buildingarea > landsize: 311


In [84]:
print(f"Number of entries with Bedroom2 > Rooms: {(data_raw_num['Bedroom2'] > data_raw_num['Rooms']).sum()}")
print(f"Number of entries with Bathroom > Rooms: {(data_raw_num['Bathroom'] > data_raw_num['Rooms']).sum()}")

Number of entries with Bedroom2 > Rooms: 221
Number of entries with Bathroom > Rooms: 36


In [103]:
data_raw_num.loc[data_raw_num['Bedroom2'] > data_raw_num['Rooms'], 'Bedroom2'] = None

### 2.2 Categorical data

In [91]:
data_raw_cat.nunique()

Suburb           345
Address        26751
Type               3
Method             5
SellerG          349
Date              78
CouncilArea       33
Regionname         8
dtype: int64

In [ ]:
# The address is unique for each entry, so it is not useful for a model.
data_raw_cat.drop(columns=['Address'], inplace=True)

In [99]:
data_raw_cat["Type"].value_counts()

Type
h    18472
u     5909
t     2866
Name: count, dtype: int64

In [94]:
data_raw_cat["Method"].value_counts()

Method
S     17515
SP     3603
PI     3255
VB     2684
SA      190
Name: count, dtype: int64

In [98]:
data_raw_cat["SellerG"].value_counts()

SellerG
Nelson           2735
Jellis           2532
Barry            2393
hockingstuart    2110
Ray              1581
                 ... 
Craig               1
Reed                1
White               1
Oak                 1
Sanctuary           1
Name: count, Length: 349, dtype: int64

In [100]:
data_raw_cat["Date"].value_counts()

Date
28/10/2017    879
17/03/2018    753
24/02/2018    723
9/12/2017     723
25/11/2017    682
             ... 
4/02/2016      35
30/09/2017     18
20/01/2018     12
27/01/2018     11
28/01/2016      2
Name: count, Length: 78, dtype: int64

In [106]:
data_raw_cat["Date"].max()

'9/12/2017'

In [101]:
data_raw_cat["CouncilArea"].value_counts()

CouncilArea
Boroondara City Council           2520
Darebin City Council              2349
Moreland City Council             1790
Glen Eira City Council            1643
Moonee Valley City Council        1584
Melbourne City Council            1502
Banyule City Council              1457
Brimbank City Council             1366
Bayside City Council              1311
Maribyrnong City Council          1221
Hume City Council                 1036
Monash City Council               1007
Port Phillip City Council          952
Yarra City Council                 918
Stonnington City Council           884
Manningham City Council            842
Hobsons Bay City Council           799
Kingston City Council              735
Whittlesea City Council            709
Wyndham City Council               492
Whitehorse City Council            435
Maroondah City Council             356
Knox City Council                  287
Melton City Council                243
Frankston City Council             229
Greater Dande

In [102]:
data_raw_cat["Regionname"].value_counts()

Regionname
Southern Metropolitan         8524
Northern Metropolitan         7864
Western Metropolitan          5815
Eastern Metropolitan          3272
South-Eastern Metropolitan    1341
Eastern Victoria               166
Northern Victoria              166
Western Victoria                96
Name: count, dtype: int64

# 3. Splitting of data into training and test data